In [ ]:
import numpy as np
import pandas as pd

# From https://towardsdatascience.com/reverse-geocoding-in-python-a915acf29eb6
# %load_ext autotime
import geopandas as gpd
import geopy
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter
import matplotlib.pyplot as plt
import plotly_express as px
import tqdm
from tqdm.notebook import tqdm_notebook

# From https://geocoder.readthedocs.io/
import geocoder

import requests
import io
import os

import datetime as dt
from datetime import date
from datetime import timedelta

from copy import deepcopy

pd.options.mode.chained_assignment = None  # default='warn'

### 1. Read TaxData.csv and do some EDA

In [ ]:
tax_data_df = pd.read_csv("TaxData.csv", low_memory=False)

In [7]:
tax_data_df.head()

ORIGINAL APN  CENSUS TRACT COUNTY USE1 ZONING MUNICIPALITY NAME  \
0  30-017-02-0077-002  5.765002e+09         364   R 1B        ALLEN PARK   
1  30-017-02-0079-002  5.765002e+09         364   R 1B        ALLEN PARK   
2  30-017-02-0084-002  5.766001e+09         364   R 1B        ALLEN PARK   
3  30-017-02-0086-002  5.766001e+09         364   R 1B        ALLEN PARK   
4  30-017-02-0088-002  5.766001e+09         364   R 1B        ALLEN PARK   

  MUNICIPALITY CODE  PROPERTY LEVEL LATITUDE  PROPERTY LEVEL LONGITUDE  ACRES  \
0                30                42.243229                -83.205078  0.116   
1                30                42.243231                -83.204885  0.117   
2                30                42.242630                -83.204984  0.116   
3                30                42.242630                -83.205097  0.116   
4                30                42.242631                -83.205245  0.116   

   YEAR BUILT  LIVING SQUARE FEET  NUMBER OF UNITS  
0      1950.0              1031.0              1.0  
1      1950.0              1356.0              1.0  
2      1950.0              1031.0              1.0  
3      1950.0              1031.0              1.0  
4      1950.0              1031.0              1.0

In [6]:
tax_data_df.describe()

CENSUS TRACT  PROPERTY LEVEL LATITUDE  PROPERTY LEVEL LONGITUDE  \
count  4.993588e+06             4.998132e+06              4.998132e+06   
mean   3.008229e+09             4.310223e+01             -8.436333e+01   
std    3.396698e+09             1.103893e+00              1.318532e+00   
min    1.000000e+02             2.905371e+01             -9.713022e+01   
25%    5.100204e+07             4.238418e+01             -8.543257e+01   
50%    1.578003e+09             4.268700e+01             -8.394964e+01   
75%    5.515001e+09             4.343342e+01             -8.325475e+01   
max    9.859001e+09             4.747659e+01             -8.241986e+01   

              ACRES    YEAR BUILT  LIVING SQUARE FEET  NUMBER OF UNITS  
count  4.530966e+06  1.991960e+06        1.982485e+06     2.167832e+06  
mean   2.409751e+01  1.959695e+03        1.696460e+03     1.027096e+00  
std    5.048321e+03  2.857208e+01        4.409584e+03     3.778352e-01  
min    1.000000e-04  1.700000e+03        1.000000e+00     1.000000e+00  
25%    1.650000e-01  1.942000e+03        1.064000e+03     1.000000e+00  
50%    3.630000e-01  1.958000e+03        1.440000e+03     1.000000e+00  
75%    2.000000e+00  1.980000e+03        1.998000e+03     1.000000e+00  
max    3.102156e+06  2.016000e+03        4.622108e+06     9.100000e+01

In [5]:
tax_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5070108 entries, 0 to 5070107
Data columns (total 12 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   ORIGINAL APN              object 
 1   CENSUS TRACT              float64
 2   COUNTY USE1               object 
 3   ZONING                    object 
 4   MUNICIPALITY NAME         object 
 5   MUNICIPALITY CODE         object 
 6   PROPERTY LEVEL LATITUDE   float64
 7   PROPERTY LEVEL LONGITUDE  float64
 8   ACRES                     float64
 9   YEAR BUILT                float64
 10  LIVING SQUARE FEET        float64
 11  NUMBER OF UNITS           float64
dtypes: float64(7), object(5)
memory usage: 464.2+ MB


In [8]:
tax_data_df = tax_data_df.rename(columns={"PROPERTY LEVEL LATITUDE": "LAT", 
                                          "PROPERTY LEVEL LONGITUDE": "LNG"})
tax_data_df["AGE"] = dt.date.today().year - tax_data_df["YEAR BUILT"].iloc[0]
tax_data_df["PLACE NAME"] = ""
tax_data_df["ZIP CODE"] = ""

In [9]:
tax_data_df = tax_data_df[tax_data_df["LAT"].notnull()]

In [10]:
tax_data_df.isnull().sum()

ORIGINAL APN                0
CENSUS TRACT            47243
COUNTY USE1                 0
ZONING                3791388
MUNICIPALITY NAME           2
MUNICIPALITY CODE           2
LAT                         0
LNG                         0
ACRES                  518616
YEAR BUILT            3006384
LIVING SQUARE FEET    3015723
NUMBER OF UNITS       2831891
AGE                         0
PLACE NAME                  0
ZIP CODE                    0
dtype: int64

In [11]:
tax_data_df.head()

ORIGINAL APN  CENSUS TRACT COUNTY USE1 ZONING MUNICIPALITY NAME  \
0  30-017-02-0077-002  5.765002e+09         364   R 1B        ALLEN PARK   
1  30-017-02-0079-002  5.765002e+09         364   R 1B        ALLEN PARK   
2  30-017-02-0084-002  5.766001e+09         364   R 1B        ALLEN PARK   
3  30-017-02-0086-002  5.766001e+09         364   R 1B        ALLEN PARK   
4  30-017-02-0088-002  5.766001e+09         364   R 1B        ALLEN PARK   

  MUNICIPALITY CODE        LAT        LNG  ACRES  YEAR BUILT  \
0                30  42.243229 -83.205078  0.116      1950.0   
1                30  42.243231 -83.204885  0.117      1950.0   
2                30  42.242630 -83.204984  0.116      1950.0   
3                30  42.242630 -83.205097  0.116      1950.0   
4                30  42.242631 -83.205245  0.116      1950.0   

   LIVING SQUARE FEET  NUMBER OF UNITS   AGE PLACE NAME ZIP CODE  
0              1031.0              1.0  70.0                      
1              1356.0              1.0  70.0                      
2              1031.0              1.0  70.0                      
3              1031.0              1.0  70.0                      
4              1031.0              1.0  70.0

In [94]:
len(tax_data_df)

4998132

### 1.1 Sample lat/lng coordinates to test reverse geocoding

In [92]:
# Lat and lng of Googleplex
lat1 = "37.419857"
lng1 = "-122.078827"

In [93]:
# Lat and lng of first entry in tax_data_df
lat2 = "42.243229"
lng2 = "-83.205078"

## 2. Subset tax_data_df to test geocoding

In [33]:
# Data set is too large to perform API call in a single for loop
tax_data_df_1 = tax_data_df.iloc[0:10]
tax_data_df_1

ORIGINAL APN  CENSUS TRACT COUNTY USE1 ZONING MUNICIPALITY NAME  \
0  30-017-02-0077-002  5.765002e+09         364   R 1B        ALLEN PARK   
1  30-017-02-0079-002  5.765002e+09         364   R 1B        ALLEN PARK   
2  30-017-02-0084-002  5.766001e+09         364   R 1B        ALLEN PARK   
3  30-017-02-0086-002  5.766001e+09         364   R 1B        ALLEN PARK   
4  30-017-02-0088-002  5.766001e+09         364   R 1B        ALLEN PARK   
5  30-017-02-0090-002  5.766001e+09         364   R 1B        ALLEN PARK   
6  30-017-02-0093-000  5.766001e+09         364   R 1B        ALLEN PARK   
7  30-017-02-0095-000  5.766001e+09         364   R 1B        ALLEN PARK   
8  30-017-02-0097-000  5.766001e+09         364   R 1B        ALLEN PARK   
9  30-017-02-0101-002  5.766001e+09         364   R 1B        ALLEN PARK   

  MUNICIPALITY CODE        LAT        LNG  ACRES  YEAR BUILT  \
0                30  42.243229 -83.205078  0.116      1950.0   
1                30  42.243231 -83.204885  0.117      1950.0   
2                30  42.242630 -83.204984  0.116      1950.0   
3                30  42.242630 -83.205097  0.116      1950.0   
4                30  42.242631 -83.205245  0.116      1950.0   
5                30  42.242628 -83.205442  0.114      1950.0   
6                30  42.242626 -83.205950  0.114      1956.0   
7                30  42.242624 -83.206113  0.101      1951.0   
8                30  42.242624 -83.206268  0.101      1956.0   
9                30  42.242623 -83.206635  0.139      1956.0   

   LIVING SQUARE FEET  NUMBER OF UNITS   AGE  \
0              1031.0              1.0  70.0   
1              1356.0              1.0  70.0   
2              1031.0              1.0  70.0   
3              1031.0              1.0  70.0   
4              1031.0              1.0  70.0   
5              1443.0              1.0  70.0   
6              1570.0              1.0  70.0   
7              1031.0              1.0  70.0   
8              1047.0              1.0  70.0   
9              1052.0              1.0  70.0   

                                          PLACE NAME ZIP CODE  
0  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
1  Vine Avenue, Allen Park, Wayne County, Michiga...    48101  
2  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
3  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
4  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
5  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
6  Rosedale Boulevard, Allen Park, Wayne County, ...    48101  
7  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
8  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
9  Hubert Avenue, Allen Park, Wayne County, Michi...    48101

To reverse geocode the entire dataset (4998132 rows with non-null lat/lon), we can employ MapReduce or simply break into chunks and then concatenate the results

## 3. Reverse geocoding using free Openstreetmap geocoding service

### 3.1 Inspect API call result using sample lat/lng

In [95]:
# Googleplex
coordinates = lat1, lng1
locator = Nominatim(user_agent="myGeocoder")
location = locator.reverse(coordinates)
result = location.raw
result

{'place_id': 118936921,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': 125372590,
 'lat': '37.41984925',
 'lon': '-122.07885167838052',
 'display_name': 'Google 1842, 1842, North Shoreline Boulevard, Mountain View, Santa Clara County, California, 94043, United States',
 'address': {'office': 'Google 1842',
  'house_number': '1842',
  'road': 'North Shoreline Boulevard',
  'city': 'Mountain View',
  'county': 'Santa Clara County',
  'state': 'California',
  'postcode': '94043',
  'country': 'United States',
  'country_code': 'us'},
 'boundingbox': ['37.4195758', '37.4201344', '-122.0791846', '-122.0785385']}

In [96]:
# First entry in tax_data_df
coordinates = lat2, lng2
locator = Nominatim(user_agent="myGeocoder")
location = locator.reverse(coordinates)
result = location.raw
result

{'place_id': 85975463,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': 8771706,
 'lat': '42.24299907114748',
 'lon': '-83.20507697983786',
 'display_name': 'Wick Road, Allen Park, Wayne County, Michigan, 48101, United States',
 'address': {'road': 'Wick Road',
  'town': 'Allen Park',
  'county': 'Wayne County',
  'state': 'Michigan',
  'postcode': '48101',
  'country': 'United States',
  'country_code': 'us'},
 'boundingbox': ['42.242862', '42.243001', '-83.208981', '-83.204641']}

### 3.1 Create function to make Openstreetma API calls

In [83]:
def rev_geocoder_osm(df):
    
    for i in range(len(df)):
        lat = df.iloc[i]["LAT"]
        lng = df.iloc[i]["LNG"]
        coordinates = lat, lng
        
        locator = Nominatim(user_agent="myGeocoder")
        location = locator.reverse(coordinates)
        result = location.raw
        
        df["PLACE NAME"].iloc[i] = result["display_name"]
        df["ZIP CODE"].iloc[i] = result["address"]["postcode"]
        
    return df

In [84]:
df_osm = rev_geocoder_osm(tax_data_df_1)

In [85]:
df_osm

ORIGINAL APN  CENSUS TRACT COUNTY USE1 ZONING MUNICIPALITY NAME  \
0  30-017-02-0077-002  5.765002e+09         364   R 1B        ALLEN PARK   
1  30-017-02-0079-002  5.765002e+09         364   R 1B        ALLEN PARK   
2  30-017-02-0084-002  5.766001e+09         364   R 1B        ALLEN PARK   
3  30-017-02-0086-002  5.766001e+09         364   R 1B        ALLEN PARK   
4  30-017-02-0088-002  5.766001e+09         364   R 1B        ALLEN PARK   
5  30-017-02-0090-002  5.766001e+09         364   R 1B        ALLEN PARK   
6  30-017-02-0093-000  5.766001e+09         364   R 1B        ALLEN PARK   
7  30-017-02-0095-000  5.766001e+09         364   R 1B        ALLEN PARK   
8  30-017-02-0097-000  5.766001e+09         364   R 1B        ALLEN PARK   
9  30-017-02-0101-002  5.766001e+09         364   R 1B        ALLEN PARK   

  MUNICIPALITY CODE        LAT        LNG  ACRES  YEAR BUILT  \
0                30  42.243229 -83.205078  0.116      1950.0   
1                30  42.243231 -83.204885  0.117      1950.0   
2                30  42.242630 -83.204984  0.116      1950.0   
3                30  42.242630 -83.205097  0.116      1950.0   
4                30  42.242631 -83.205245  0.116      1950.0   
5                30  42.242628 -83.205442  0.114      1950.0   
6                30  42.242626 -83.205950  0.114      1956.0   
7                30  42.242624 -83.206113  0.101      1951.0   
8                30  42.242624 -83.206268  0.101      1956.0   
9                30  42.242623 -83.206635  0.139      1956.0   

   LIVING SQUARE FEET  NUMBER OF UNITS   AGE  \
0              1031.0              1.0  70.0   
1              1356.0              1.0  70.0   
2              1031.0              1.0  70.0   
3              1031.0              1.0  70.0   
4              1031.0              1.0  70.0   
5              1443.0              1.0  70.0   
6              1570.0              1.0  70.0   
7              1031.0              1.0  70.0   
8              1047.0              1.0  70.0   
9              1052.0              1.0  70.0   

                                          PLACE NAME ZIP CODE  
0  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
1  Vine Avenue, Allen Park, Wayne County, Michiga...    48101  
2  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
3  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
4  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
5  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
6  Rosedale Boulevard, Allen Park, Wayne County, ...    48101  
7  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
8  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
9  Hubert Avenue, Allen Park, Wayne County, Michi...    48101

## 4. Reverse geocoding using Google Geocoding API

### 4.1 Inspect API call result using sample lat/lng  
Note: This returns a very long dict

In [104]:
# Googleplex
lat = lat1
lng = lng1

# Set URL string
url = "https://maps.googleapis.com/maps/api/geocode/json?latlng="+lat+","+lng+"&key="+key

response = requests.get(url)
result = response.json()
result

{'plus_code': {'compound_code': 'CW9C+WF Mountain View, CA, USA',
  'global_code': '849VCW9C+WF'},
 'results': [{'address_components': [{'long_name': '1842',
     'short_name': '1842',
     'types': ['street_number']},
    {'long_name': 'North Shoreline Boulevard',
     'short_name': 'N Shoreline Blvd',
     'types': ['route']},
    {'long_name': 'Mountain View',
     'short_name': 'Mountain View',
     'types': ['locality', 'political']},
    {'long_name': 'Santa Clara County',
     'short_name': 'Santa Clara County',
     'types': ['administrative_area_level_2', 'political']},
    {'long_name': 'California',
     'short_name': 'CA',
     'types': ['administrative_area_level_1', 'political']},
    {'long_name': 'United States',
     'short_name': 'US',
     'types': ['country', 'political']},
    {'long_name': '94043', 'short_name': '94043', 'types': ['postal_code']}],
   'formatted_address': '1842 N Shoreline Blvd, Mountain View, CA 94043, USA',
   'geometry': {'location': {'lat': 37

In [105]:
# First entry in tax_data_df 
lat = lat2 
lng = lng2

# Set URL string
url = "https://maps.googleapis.com/maps/api/geocode/json?latlng="+lat+","+lng+"&key="+key

response = requests.get(url)
result = response.json()
result

{'plus_code': {'compound_code': '6QVV+7X Allen Park, MI, USA',
  'global_code': '86JR6QVV+7X'},
 'results': [{'address_components': [{'long_name': '4714',
     'short_name': '4714',
     'types': ['street_number']},
    {'long_name': 'Wick Road', 'short_name': 'Wick Rd', 'types': ['route']},
    {'long_name': 'Allen Park',
     'short_name': 'Allen Park',
     'types': ['locality', 'political']},
    {'long_name': 'Wayne County',
     'short_name': 'Wayne County',
     'types': ['administrative_area_level_2', 'political']},
    {'long_name': 'Michigan',
     'short_name': 'MI',
     'types': ['administrative_area_level_1', 'political']},
    {'long_name': 'United States',
     'short_name': 'US',
     'types': ['country', 'political']},
    {'long_name': '48101', 'short_name': '48101', 'types': ['postal_code']}],
   'formatted_address': '4714 Wick Rd, Allen Park, MI 48101, USA',
   'geometry': {'bounds': {'northeast': {'lat': 42.2432658,
      'lng': -83.2049682},
     'southwest': {'l

### 4.2 Explore result of API call

In [71]:
result.get("results")[0]

{'address_components': [{'long_name': '4714',
   'short_name': '4714',
   'types': ['street_number']},
  {'long_name': 'Wick Road', 'short_name': 'Wick Rd', 'types': ['route']},
  {'long_name': 'Allen Park',
   'short_name': 'Allen Park',
   'types': ['locality', 'political']},
  {'long_name': 'Wayne County',
   'short_name': 'Wayne County',
   'types': ['administrative_area_level_2', 'political']},
  {'long_name': 'Michigan',
   'short_name': 'MI',
   'types': ['administrative_area_level_1', 'political']},
  {'long_name': 'United States',
   'short_name': 'US',
   'types': ['country', 'political']},
  {'long_name': '48101', 'short_name': '48101', 'types': ['postal_code']}],
 'formatted_address': '4714 Wick Rd, Allen Park, MI 48101, USA',
 'geometry': {'bounds': {'northeast': {'lat': 42.2432658, 'lng': -83.2049682},
   'southwest': {'lat': 42.2431537, 'lng': -83.2050752}},
  'location': {'lat': 42.2432065, 'lng': -83.20502359999999},
  'location_type': 'ROOFTOP',
  'viewport': {'northe

In [78]:
# Get formatted address from result
result.get("results")[0].get("formatted_address")

'4714 Wick Rd, Allen Park, MI 48101, USA'

In [77]:
result.get("results")[0].get("address_components")

[{'long_name': '4714', 'short_name': '4714', 'types': ['street_number']},
 {'long_name': 'Wick Road', 'short_name': 'Wick Rd', 'types': ['route']},
 {'long_name': 'Allen Park',
  'short_name': 'Allen Park',
  'types': ['locality', 'political']},
 {'long_name': 'Wayne County',
  'short_name': 'Wayne County',
  'types': ['administrative_area_level_2', 'political']},
 {'long_name': 'Michigan',
  'short_name': 'MI',
  'types': ['administrative_area_level_1', 'political']},
 {'long_name': 'United States',
  'short_name': 'US',
  'types': ['country', 'political']},
 {'long_name': '48101', 'short_name': '48101', 'types': ['postal_code']}]

In [82]:
# Get zip code from result
result.get("results")[0].get("address_components")[6]["long_name"]

'48101'

### 4.3 Create function to make Google Geocoding API calls

In [86]:
def rev_geocoder_google():

    for i in range(len(df)):
        lat = df.iloc[i]["LAT"]
        lng = df.iloc[i]["LNG"]
        
        key="AIzaSyD-NRUsF_OkkWaaz_iWootCLYlp2UR6hPw"
        url = "https://maps.googleapis.com/maps/api/geocode/json?latlng="+lat+","+lng+"&key="+key
        
        response = requests.get(url)
        result = response.json()

        df["PLACE NAME"].iloc[i] = result.get("results")[0].get("address_components")
        df["ZIP CODE"].iloc[i] = result.get("results")[0].get("address_components")[6]["long_name"]
        
    return df

In [87]:
df_google = rev_geocoder_osm(tax_data_df_1)

In [88]:
df_google

ORIGINAL APN  CENSUS TRACT COUNTY USE1 ZONING MUNICIPALITY NAME  \
0  30-017-02-0077-002  5.765002e+09         364   R 1B        ALLEN PARK   
1  30-017-02-0079-002  5.765002e+09         364   R 1B        ALLEN PARK   
2  30-017-02-0084-002  5.766001e+09         364   R 1B        ALLEN PARK   
3  30-017-02-0086-002  5.766001e+09         364   R 1B        ALLEN PARK   
4  30-017-02-0088-002  5.766001e+09         364   R 1B        ALLEN PARK   
5  30-017-02-0090-002  5.766001e+09         364   R 1B        ALLEN PARK   
6  30-017-02-0093-000  5.766001e+09         364   R 1B        ALLEN PARK   
7  30-017-02-0095-000  5.766001e+09         364   R 1B        ALLEN PARK   
8  30-017-02-0097-000  5.766001e+09         364   R 1B        ALLEN PARK   
9  30-017-02-0101-002  5.766001e+09         364   R 1B        ALLEN PARK   

  MUNICIPALITY CODE        LAT        LNG  ACRES  YEAR BUILT  \
0                30  42.243229 -83.205078  0.116      1950.0   
1                30  42.243231 -83.204885  0.117      1950.0   
2                30  42.242630 -83.204984  0.116      1950.0   
3                30  42.242630 -83.205097  0.116      1950.0   
4                30  42.242631 -83.205245  0.116      1950.0   
5                30  42.242628 -83.205442  0.114      1950.0   
6                30  42.242626 -83.205950  0.114      1956.0   
7                30  42.242624 -83.206113  0.101      1951.0   
8                30  42.242624 -83.206268  0.101      1956.0   
9                30  42.242623 -83.206635  0.139      1956.0   

   LIVING SQUARE FEET  NUMBER OF UNITS   AGE  \
0              1031.0              1.0  70.0   
1              1356.0              1.0  70.0   
2              1031.0              1.0  70.0   
3              1031.0              1.0  70.0   
4              1031.0              1.0  70.0   
5              1443.0              1.0  70.0   
6              1570.0              1.0  70.0   
7              1031.0              1.0  70.0   
8              1047.0              1.0  70.0   
9              1052.0              1.0  70.0   

                                          PLACE NAME ZIP CODE  
0  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
1  Vine Avenue, Allen Park, Wayne County, Michiga...    48101  
2  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
3  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
4  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
5  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
6  Rosedale Boulevard, Allen Park, Wayne County, ...    48101  
7  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
8  Wick Road, Allen Park, Wayne County, Michigan,...    48101  
9  Hubert Avenue, Allen Park, Wayne County, Michi...    48101

Attempts to run either of the two functions on the entire dataset were unsuccessful and ended with the notebook hanging and requiring the kernel to be restarted